### PR TITLE
Add API Gateway (proxy level)

### PR DESCRIPTION
**Pull request by huynhminhtufu**
The **API gateway** is responsible for **dispatch** the requests to the **corresponding** REST endpoint (reverse proxy), load balancing, simple heartbeat check as well as failure handling (with circuit breaker).

### CODE REVIEW

**Comment by huynhminhtufu** [comment]
@licolsan Please help me to improve this code (select service), maybe use CONST + ENUM?

image

**Comment by huynhminhtufu** [comment]
Example Model: 
image

File: src/main/java/software/pet/api/http/APIGateway.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        });
+        cRes.endHandler((v) -> req.response().end());
+      });
+      cReq.setChunked(true);
+      cReq.headers().setAll(req.headers());
+      req.handler(data -> {
+        cReq.write(data);
+      });
+      req.endHandler((v) -> cReq.end());
+    }).listen(5000);
```
Comment (licolsan) [inline src/main/java/software/pet/api/http/APIGateway.java]: Why not `listen(servicePort)`

File: src/main/java/software/pet/api/http/APIGateway.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        JsonObject responseJson = new JsonObject().put("error", "Not found");
+        req.response().putHeader("content-type", "application/json; charset=utf-8")
+            .end(Json.encodePrettily(responseJson));
+        return;
+      }
+
+      String serviceName = path.substring(initialOffset).split("/")[1];
+      String newPath = pathLength >= 3 ? path.substring(initialOffset).split("/")[2] : "";
+
+      int servicePort = 5000;
```
Comment (licolsan) [inline src/main/java/software/pet/api/http/APIGateway.java]: You can put it in `env` file

File: src/main/java/software/pet/api/user/UserVerticle.java [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
 import io.vertx.reactivex.core.AbstractVerticle;
 import io.vertx.serviceproxy.ServiceBinder;
+import software.pet.api.user.api.UserHandler;
+import software.pet.api.user.api.UserRouter;
 
 public class UserVerticle extends AbstractVerticle {
+  final static Logger logger = LoggerFactory.getLogger(UserVerticle.class);
 
   public static final String USER_SERVICE_ADDRESS = "user-service";
+  public static final int USER_API_PORT = 5002;
```
Comment (licolsan) [inline src/main/java/software/pet/api/user/UserVerticle.java]: You can put it in `env` file

File: src/main/java/software/pet/api/http/APIGateway.java (hunk 2 above)
Comment (licolsan) [inline src/main/java/software/pet/api/http/APIGateway.java]: You can put it in env file

File: src/main/java/software/pet/api/user/UserVerticle.java (hunk 3 above)
Comment (licolsan) [inline src/main/java/software/pet/api/user/UserVerticle.java]: You can put it in env file

File: src/main/java/software/pet/api/http/APIGateway.java (hunk 1 above)
Comment (licolsan) [inline src/main/java/software/pet/api/http/APIGateway.java]: Why not `listen(servicePort)`

**Pull request merged**